Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp

```diff
@@ -0,0 +1,338 @@
+/*
+ * Copyright (c) 2023, Nicolas Ramz <nicolas.ramz@gmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <AK/Debug.h>
+#include <AK/Endian.h>
+#include <AK/FixedArray.h>
+#include <LibGfx/ImageFormats/ILBMLoader.h>
+
+namespace Gfx {
+
+struct IFF_FILE_HEADER {
+    FourCC form;
+    BigEndian<u32> file_size;
+    FourCC format;
+};
+
+static_assert(AssertSize<IFF_FILE_HEADER, 12>());
+
+struct Chunk {
+    FourCC type;
+    ReadonlyBytes data;
+};
+
+enum class CompressionType : u8 {
+    None = 0,
+    ByteRun = 1
+};
+
+enum class MaskType : u8 {
+    None = 0,
+    HasMask = 1,
+    HasTransparentColor = 2,
+    HasLasso = 3
+};
+
+struct ChunkHeader {
+    FourCC chunk_type;
+    BigEndian<u32> chunk_size;
+};
+
+struct BMHDHeader {
+    BigEndian<u16> width;
+    BigEndian<u16> height;
+    BigEndian<i16> x;
+    BigEndian<i16> y;
+    u8 planes;
+    MaskType mask;
+    CompressionType compression;
+    u8 pad;
+    BigEndian<u16> transparent_color;
+    u8 x_aspect;
+    u8 y_aspect;
+    BigEndian<u16> page_width;
+    BigEndian<u16> page_height;
+};
+
+static_assert(sizeof(BMHDHeader) == 20);
+
+struct ILBMLoadingContext {
+    enum class State {
+        NotDecoded = 0,
+        HeaderDecoded,
+        BitmapHeaderDecoded,
+        BitmapDecoded
+    };
+    State state { State::NotDecoded };
+    ReadonlyBytes data;
+
+    // points to current chunk
+    ReadonlyBytes chunks_cursor;
+
+    // max number of bytes per plane row
+    u16 pitch;
+
+    FixedArray<Color> color_table;
+
+    RefPtr<Gfx::Bitmap> bitmap;
+
+    BMHDHeader bm_header;
+};
+
+static ErrorOr<void> decode_iff_ilbm_header(ILBMLoadingContext& context)
+{
+    if (context.state >= ILBMLoadingContext::State::HeaderDecoded)
+        return {};
+
+    if (context.data.size() < sizeof(IFF_FILE_HEADER))
+        return Error::from_string_literal("Missing IFF header");
+
+    auto& header = *bit_cast<IFF_FILE_HEADER const*>(context.data.data());
+    if (header.form != FourCC("FORM") || header.format != FourCC("ILBM"))
+        return Error::from_string_literal("Invalid IFF-ILBM header");
+
+    return {};
+}
+
+static ErrorOr<FixedArray<Color>> decode_cmap_chunk(Chunk cmap_chunk)
+{
+    auto& data = cmap_chunk.data;
+    u8 const size = data.size() / 3;
+    FixedArray<Color> color_table = TRY(FixedArray<Color>::create(size));
+
+    for (int i = 0; i < size; ++i) {
+        color_table[i] = Color(data[i * 3], data[(i * 3) + 1], data[(i * 3) + 2]);
+    }
+
+    return color_table;
+}
+
+static ErrorOr<RefPtr<Gfx::Bitmap>> chunky_to_bitmap(ILBMLoadingContext& context, ByteBuffer const& chunky)
+{
+    auto& width = context.bm_header.width;
```

Comment:
The basic rule of thumb for should I take a reference or a copy to an object is: if it fits into a register a copy is cheaper, take a reference otherwise.
Integers fall into the first category, don't take a reference if it's not needed.

Also, when I see someone taking a mutable reference, I expect it to be used to be written to. Explicitly taking a `const` reference makes your intentions much clearer.